wagtail/admin/panels.py

```diff
@@ -340,6 +342,12 @@ def is_shown(self):
             """
             return True
 
+        def is_custom_rendered(self):
+            """
+            Whether this panel has custom rendering that makes it visually separate from the rest of the form.
+            """
+            return False
```

Comment:
Not a big fan of this honestly, but I can't think of a cleaner way to do this. I needed a way to make `TabbedInterface` render without a tab if all of its contents are "custom-rendered" (i.e. rendered outside the tab's content area).

Imagine:
```
TabbedInterface(
    children=[
        ObjectList("Content", children=[...]),
        ObjectList("Promote", children=[...]),
        ObjectList("Settings", children=[
            PublishingPanel(), # is custom-rendered
            CommentPanel(), # is custom-rendered
            ... # user-defined panels (if any), that may not be custom-rendered
    ]
)
```

To determine if the Settings tab should be rendered with the tab header or just the children, we need to:
- Add a flag to `PublishingPanel` and `CommentPanel` (and possibly others in the future) that indicates "I'm rendered outside of the tab's content section -- 'ignore' me, but please still render my markup!"
- Have a way to retrieve this information from the `TabbedInterface`. It needs to go through the `ObjectList`s and check the flag for all of its children. If all of the children has the flag set, then the tab header should not be rendered. Otherwise, render the tab normally. (Now that I think about this, this could be slightly optimised by checking if there's at least one panel with the flag unset).

Reply:
This feels unideal but at least it follows the existing `is_shown` logic, though in this case I think the two panels could be in another tab altogether, and our ObjectList / panel could have an "is empty" check to panels similarly so it skips rendering. Perhaps even in reusing the `is_shown` for this purpose.

On naming, to me `is_custom_rendered` is too generic – if we retained this approach, it’d be nice for it to be `is_visible` or `has_visible_children`, similarly to the existing `visible_children`. Perhaps rename `is_shown` so it’s clear one is about visual rendering and the other about semantics.

Reply:
It's really hard for me to find the most suitable naming for this. Currently, we have:

- `is_shown()`: [documented]((https://docs.wagtail.org/en/latest/reference/panel_api.html#wagtail.admin.panels.Panel.BoundPanel.is_shown)) ("Whether this panel should be rendered; if false, it is skipped in the template output."
- `visible_children` and `visible_children_with_identifiers`: not documented, used in templates to skip the rendering entirely for children that have `is_shown()` returning `False`.

If we have both `is_shown` and `is_visible`, which one should be used in `visible_children`? To me, `is_visible` seems to make more sense given the words are similar. But that means we're renaming `is_shown` to `is_visible`, and there might be people out there who has overridden `is_shown`, making this a breaking change. If we leave `is_shown()` as-is and introduce `is_visible`, people might assume it's used in `visible_children`. There might be other words e.g. `is_displayed`, but I don't think that solves the problem...

For now, I've changed `is_custom_rendered` to `is_shown_immediately` (with the logic negated). The method indicates whether the panel is shown immediately (e.g. instead of shown in a modal). Suggestions welcome!